### PR TITLE
Retool required field functionality

### PR DESCRIPTION
**Pull request by jrbotros**
Genericizes field validation to use the HTML elements themselves (rather than their ng-model attribute) to support fields nested inside directives. Also allows defining custom field types and registering corresponding validator functions.

### CODE REVIEW

File: orchestra/static/orchestra/task/js/required_fields.js [hunk 1]
```diff
@@ -9,45 +9,73 @@
...
+      require: function(fieldType, field) {
         /**
-         * Sets a field as required. Fields are specified by
-         * dot-delimited key strings (e.g., `key0.key1.key2`).
+         * Sets a field as required. Fields are HTML elements to be
+         * checked by one or more validators according to their field
+         * type.
          */
-        this.fields = this.fields.concat(fields)
+        if (Object.keys(this.fields).indexOf(fieldType) >= 0) {
```
Comment (thisisdhaas) [inline orchestra/static/orchestra/task/js/required_fields.js]: is this the standard way to check for a key on an associative array? Doesn't just `if (this.fields[fieldType] != undefined)` work?

File: orchestra/static/orchestra/task/js/required_fields.js [hunk 2]
```diff
@@ -9,45 +9,73 @@
...
-              requiredFields.invalid.push(field)
-              break;
+        for (var fieldType in requiredFields.fields) {
+          var validators = requiredFields.validators[fieldType];
+          var fields = requiredFields.fields[fieldType];
+          fields.forEach(function(field) {
+            var success = true;
+            validators.forEach(function(validator) {
+              success = success && validator(field);
+            })
```
Comment (thisisdhaas) [inline orchestra/static/orchestra/task/js/required_fields.js:63]: should we catch the case where someone requires validation for a non-existent field type? e.g., fields contains `'input-radio'` but we don't support it? Here, validators will be undefined and `forEach` will barf confusingly, I think.

File: orchestra/static/orchestra/task/js/required_fields.js [hunk 3]
```diff
@@ -9,45 +9,73 @@
...
+        }
         $rootScope.$broadcast('orchestra:task:validatedFields');
         return requiredFields.invalid.length === 0;
+      },
+      registerValidator: function(fieldType, validator) {
+        /**
+         * Register a validator function to the given field type.
+         */
+        var requiredFields = this;
+        if (Object.keys(requiredFields.validators).indexOf(fieldType) >= 0) {
```
Comment (thisisdhaas) [inline orchestra/static/orchestra/task/js/required_fields.js]: same comment.

File: orchestra/static/orchestra/task/js/required_fields.js [hunk 4]
```diff
@@ -9,45 +9,73 @@
...
+          validator[fieldType].push(validator);
```
Comment (thisisdhaas) [inline orchestra/static/orchestra/task/js/required_fields.js]: `validator`? Or `requiredFields.validators`?

File: orchestra/static/orchestra/task/js/required_fields.js [hunk 5]
```diff
@@ -9,45 +9,73 @@
...
+        }
+        else {
+          validator[fieldType] = [validator];
```
Comment (thisisdhaas) [inline orchestra/static/orchestra/task/js/required_fields.js]: Same comment.

File: orchestra/static/orchestra/task/js/required_fields.js (hunk 2 above)
Comment (jrbotros) [inline orchestra/static/orchestra/task/js/required_fields.js:63]: :white_check_mark:

**Comment by thisisdhaas** [comment]
LGTM after a squash/rebase.

**Pull request merged**